cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from cirq import linalg, ops
+import numpy as np
+
+
+def _unitary_power(U, p):
+    """Raises unitary matrix U to power p."""
+    assert linalg.is_unitary(U)
+    eig_vals, eig_vectors = np.linalg.eig(U)
+    Q = np.array(eig_vectors)
+    return Q @ np.diag(np.exp(p * 1j * np.angle(eig_vals))) @ Q.conj().T
+
+
+def _is_identity(M):
+    """Checks whether M is identity."""
+    return np.allclose(M, np.eye(M.shape[0]))
+
+
+def _ry(x):
+    return np.array([[np.cos(x / 2), np.sin(x / 2)],
+                     [-np.sin(x / 2), np.cos(x / 2)]])
+
+
+def _rz(x):
+    return np.diag(np.exp([0.5j * x, -0.5j * x]))
+
+
+def _flatten(x):
+    return sum(x, [])
+
+
+class ControlledGateDecomposer:
+    """Decomposes single-qubit multi-controlled gate.
+
+    Decomposition result is a sequence of single-qubit gates,
+    CNOT gates and (optionally) CCNOT gates.
+
+    === REFERENCE ===
+    Barenco, Bennett et al. Elementary gates for quantum computation. 1995.
+    https://arxiv.org/pdf/quant-ph/9503016.pdf
+    """
+
+    def __init__(self, allow_toffoli: bool = True):
+        """
+        Args:
+            allow_toffoli - whether Toffoli (CCNOT) gates are allowed in the
+               output. If `False`, output will consist only of single-qubit
+               gates and CNOT gates. Defaults to `True`.
+        """
+        self.allow_toffoli = allow_toffoli
+
+    def _ccnot_true(self, c0: 'cirq.Qid', c1: 'cirq.Qid', target: 'cirq.Qid'):
+        """Implements action of Toffoli gate."""
+        return [
+            ops.H.on(target),
+            ops.CNOT.on(c1, target),
+            ops.ZPowGate(exponent=-0.25).on(target),
+            ops.CNOT.on(c0, target),
+            ops.T.on(target),
+            ops.CNOT.on(c1, target),
+            ops.ZPowGate(exponent=-0.25).on(target),
+            ops.CNOT.on(c0, target),
+            ops.T.on(c1),
+            ops.T.on(target),
+            ops.H.on(target),
+            ops.CNOT.on(c0, c1),
+            ops.T.on(c0),
+            ops.ZPowGate(exponent=-0.25).on(c1),
+            ops.CNOT.on(c0, c1),
+        ]
+
+    def _ccnot_congruent(self, c0: 'cirq.Qid', c1: 'cirq.Qid',
+                         target: 'cirq.Qid'):
+        """Implements 'congruent' CCNOT.
+
+        This is almost like CCNOT, but multiplies 4th qubit by -1.
+        See lemma 6.2 in [1]."""
```

Comment:
What is meant by "it multiplies the 4th qubit by -1"? There's only three qubits.

This set of operations doesn't perform a Toffoli, they perform a Toffoli with an inverted control and also there is a CCZ present.

A series of operations closer to a Toffoli is this one:

![image](https://user-images.githubusercontent.com/79941/73319069-265ff200-41f0-11ea-9cfc-c8fb71aa68cd.png)

Which is equivalent to this:

![image](https://user-images.githubusercontent.com/79941/73319078-2b24a600-41f0-11ea-8608-83a03ed9e3bd.png)

It's better because the Toffoli has no inverted controls. There is of course still some phase error in the form of the CCZ and the Z.

Reply:
Sorry, this should say "4th bit", i.e. bit 100.

About your suggestion - I'm declaring that my algorithm returns only single-qubit gates and CNOT gates. If I use cirquit your suggested - that will no longer be true.

Reply:
I kept this function but made comment clearer.